internal/telemetry/telemetry.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"encoding/json"
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/mongodb/mongocli/internal/config"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+const (
+	cacheFilename   = "telemetry"
+	dirPermissions  = 0700
+	filePermissions = 0600
+)
+
+var fs = afero.NewOsFs()
+var maxCacheFileSize int64 = 100_000_000 // 100MB
+
+type Event struct {
+	Timestamp  string                 `json:"timestamp"`
+	Source     string                 `json:"source"`
+	Name       string                 `json:"name"`
+	Properties map[string]interface{} `json:"properties"`
+}
+
+func TrackCommand(cmd *cobra.Command) {
+	if !config.TelemetryEnabled() {
+		return
+	}
+	track(cmd)
+}
+
+func track(cmd *cobra.Command) {
+	now := time.Now()
+	cmdPath := cmd.CommandPath()
+	command := strings.ReplaceAll(cmdPath, " ", "-")
+	var properties = map[string]interface{}{
+		"command": command,
+	}
+	var event = Event{
+		Timestamp:  now.Format(time.RFC3339Nano),
+		Source:     config.ToolName,
+		Name:       config.ToolName + "-event",
+		Properties: properties,
+	}
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		logError(err)
+		return
+	}
+	cacheDir = filepath.Join(cacheDir, config.ToolName)
+	err = save(event, cacheDir)
+	if err != nil {
+		logError(err)
+		return
+	}
+}
+
+func openCacheFile(cacheDir string) (afero.File, error) {
+	exists, err := afero.DirExists(fs, cacheDir)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		if mkdirError := fs.MkdirAll(cacheDir, dirPermissions); mkdirError != nil {
+			return nil, mkdirError
+		}
+	}
+	filename := filepath.Join(cacheDir, cacheFilename)
+	exists, err = afero.Exists(fs, filename)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		info, statError := fs.Stat(filename)
```

Comment:
[nit] if the file doesn't exists `fs.Stat` returns a `ErrNotFound` error (I don't remember the exact error), you can do away with the check for Exists and rely on Stat only, there's `os.IsNotExist` to do the check for you